pydantic/_internal/_typing_extra.py

```diff
@@ -315,22 +372,39 @@ def _eval_type_backport(
                 '`typing` constructs or install the `eval_type_backport` package.'
             ) from e
 
-        return eval_type_backport(value, globalns, localns, try_default=False)
+        globals_dict = globalns.resolve_namespace().unwrap_namespace() if isinstance(globalns, NsResolver) else globalns
+        locals_dict = localns.resolve_namespace().unwrap_namespace() if isinstance(localns, NsResolver) else localns
+
+        return pkg_eval_type_backport(value, globals_dict, locals_dict, try_default=False)
+
+
+def _needs_eval(value: Any) -> bool:
+    return (
+        isinstance(value, typing.ForwardRef)
+        or any(_needs_eval(arg) for arg in get_args(value))  # E.g. typing.Optional["Foo"]
+    )
 
 
 def _eval_type(
     value: Any,
-    globalns: dict[str, Any] | None = None,
-    localns: dict[str, Any] | None = None,
+    globalns: NsResolver | dict[str, Any] | None = None,
+    localns: NsResolver | dict[str, Any] | None = None,
     type_params: tuple[Any] | None = None,
 ) -> Any:
+    if not _needs_eval(value):
+        return value
+
+    # Only do resolve_namespace on demand as it can be expensive with big namespaces
+    globals_dict = globalns.resolve_namespace().unwrap_namespace() if isinstance(globalns, NsResolver) else globalns
```

Comment:
It might be possible to optimize this for the ForwardRefs case also. So we would only resolve names needed from the namespace chain as required by the value. Taking a slow path of fully merging the namespaces when that fails.